File: Information Security Lab/PRACTICE/hill_cipher.py
def get_key_matrix(plaintext):
    cols = len(plaintext) // 2
    P = [[0] * cols for _ in range(2)]
    length = 0
    while length != len(plaintext):
        for i in range(2):
            for j in range(cols):
                P[i][j] = ord(plaintext[length]) - 65
                length += 1
    return P


def get_plain_text_matrix(key):
    K = [[0] * (len(key) // 2) for _ in range(2)]
    length = 0
    for i in range(len(key) // 2):
        for j in range(2):
            K[j][i] = ord(key[length]) - 65
            length += 1
    return K


def mat_mul(matrix1, r1, c1, matrix2, r2, c2):
    ans = [[0] * c2 for _ in range(r1)]
    for i in range(r1):
        for j in range(c2):
            ans[i][j] = 0
            for k in range(c1):
                ans[i][j] += matrix1[i][k] * matrix2[k][j]
    return ans


def encrypt(plaintext, key):
    P = get_plain_text_matrix(plaintext)
    K = get_key_matrix(key)
    Encrypted = mat_mul(K, 2, len(key) // 2, P, 2, len(plaintext) // 2)

    encrypted = ""
    for i in range(len(plaintext) // 2):
        for j in range(2):
            encrypted += chr(Encrypted[j][i] % 26 + 65)

    return encrypted


def decrypt(ciphertext, key):
    C = get_plain_text_matrix(ciphertext)
    K = get_key_matrix(key)
    K_inv = [[K[1][1], -K[0][1]], [-K[1][0], K[0][0]]]
    det = K[0][0] * K[1][1] - K[0][1] * K[1][0]
    det_inv = pow(det, -1, 26)
    for i in range(2):
        for j in range(len(key) // 2):
            K_inv[i][j] = (K_inv[i][j] * det_inv) % 26
    Decrypted = mat_mul(K_inv, 2, len(key) // 2, C, 2, len(ciphertext) // 2)

    decrypted = ""
    for i in range(len(ciphertext) // 2):
        for j in range(2):
            decrypted += chr(Decrypted[j][i] % 26 + 65)

    return decrypted


plaintext = "EXAM"
key = "HILL"
ciphertext = encrypt(plaintext, key)
print("Ciphertext:", ciphertext)
decrypted_plaintext = decrypt(ciphertext, key)
print("Decrypted plaintext:", decrypted_plaintext)


"""

import numpy as np

plain_text = "ACT"
keyword = "GYBNQKURP"
alphabets = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
SIZE = 3


def modulus_mul(A, M):
    for X in range(1, M):
        if ((A % M) * (X % M)) % M == 1:
            return X
    return -1


def inverse_matrix(matrix):
    det = np.linalg.det(matrix)
    inv = np.linalg.inv(matrix)

    mul_inv = modulus_mul(int(det), 26)
    inv = (det * inv) % 26
    inv = (mul_inv * inv) % 26

    return inv


def encrypt(plain_text, keyword):
    cipher_text = ""
    message_matrix = []
    cipher_matrix = []
    key_matrix = [[0] * SIZE for i in range(SIZE)]
    temp = 0
    for char in plain_text:
        message_matrix.append([alphabets.index(char)])

    for i in range(SIZE):
        for j in range(SIZE):
            key_matrix[i][j] = alphabets.index(keyword[temp])
            temp += 1

    result = np.dot(key_matrix, message_matrix)
    cipher_matrix = result % 26

    for char in cipher_matrix:
        cipher_text += alphabets[int(char[0])]
    return cipher_text


def decrypt(cipher_text, keyword):
    cipher_matrix = []
    plain_text = ""
    key_matrix = [[0] * SIZE for i in range(SIZE)]
    temp = 0
    for char in cipher_text:
        cipher_matrix.append([alphabets.index(char)])

    for i in range(SIZE):
        for j in range(SIZE):
            key_matrix[i][j] = alphabets.index(keyword[temp])
            temp += 1

    key_matrix = inverse_matrix(key_matrix)  # Inverse of key matrix
    key_matrix = key_matrix % 26
    result = np.dot(key_matrix, cipher_matrix)
    result = np.rint(result).astype(int) % 26

    for char in result:
        plain_text += alphabets[int(char[0])]
    return plain_text


cipher_text = encrypt(plain_text, keyword)
print("Cipher text (Encrypted):", cipher_text)
print("Plain Text (Decrypted):", decrypt(cipher_text, keyword))

"""

File: Information Security Lab/PRACTICE/hill_cipher.cpp
#include <bits/stdc++.h>
using namespace std;

// Prototypes{
vector<vector<int>> matrixMultiplication(vector<vector<int>>, vector<vector<int>>);
bool isPerfectSquare(int);
vector<vector<int>> getSubmatrix(vector<vector<int>>, int, int);
int calculateDeterminant(vector<vector<int>>);
vector<vector<int>> transposeMatrix(vector<vector<int>>);
vector<vector<int>> calculateAdjoint(vector<vector<int>>);
int modInverse(int, int);
vector<vector<int>> inverseOfKeyMatrix(vector<vector<int>>);
void decrypt(string, string);
// }

int main()
{
    int choice;
    bool stop = false;
    cout << "Hill Cipher" << endl;
    while (!stop)
    {
        cout << "\n\n1. Encrypt\n2. Decrypt\n3. Exit\nEnter from your choice: ";
        cin >> choice;
        if (choice == 1)
        {
            // Encryption
            string key;
            string plainText;
            cout << "Enter key: ";
            cin >> key;
            cout << "Enter Plain Text: ";
            cin >> plainText;
            int index = sqrt(key.length());

            vector<vector<int>> keyArr(index, vector<int>(index, 0));
            vector<vector<int>> plainTextArr(index, vector<int>((int)(plainText.length() / index), 0));
            int count = 0;
            for (int i = 0; i < index; i++)
            {
                for (int j = 0; j < index; j++)
                {
                    keyArr[i][j] = (int)key[count++] - 97;
                }
            }
            count = 0;
            for (int i = 0; i < (int)(plainText.length() / index); i++)
            {
                for (int j = 0; j < index; j++)
                {
                    plainTextArr[j][i] = (int)plainText[count++] - 97;
                }
            }

            string cipherText = "";
            vector<vector<int>> mul = matrixMultiplication(keyArr, plainTextArr);
            for (int i = 0; i < mul[0].size(); i++)
            {
                for (int j = 0; j < mul.size(); j++)
                    cipherText += (char)(mul[j][i] % 26 + 97);
            }
            cout << "Cipher Text: " << cipherText << endl;
        }
        else if (choice == 2)
        {
            // Decryption
            string cipherText, key;
            cout << "Enter Key: ";
            cin >> key;
            cout << "Enter Cipher Text: ";
            cin >> cipherText;
            decrypt(cipherText, key);
        }
        else if (choice == 3)
        {
            stop = true;
        }
        else
        {
            cout << "Enter from the given choice!" << endl;
        }
    }
    return 0;
}

vector<vector<int>> matrixMultiplication(const vector<vector<int>> a, const vector<vector<int>> b)
{
    int r1 = a.size();
    int c1 = a[0].size();
    int c2 = b[0].size();
    int r2 = b.size();
    vector<vector<int>> mul(r1, vector<int>(c2, 0));

    for (int i = 0; i < r1; i++)
    {
        for (int j = 0; j < c2; j++)
        {
            for (int k = 0; k < c1; k++)
            {
                mul[i][j] += a[i][k] * b[k][j];
            }
        }
    }
    return mul;
}

bool isPerfectSquare(int a)
{
    int root = sqrt(a);
    if (root * root == a)
    {
        return true;
    }
    return false;
}

vector<vector<int>> getSubmatrix(vector<vector<int>> matrix, int rowToRemove, int colToRemove)
{
    int size = matrix.size();
    vector<vector<int>> submatrix(size - 1, vector<int>(size - 1, 0));
    int rowIndex = 0;
    int colIndex;

    for (int i = 0; i < size; i++)
    {
        if (i == rowToRemove)
        {
            continue;
        }

        colIndex = 0;
        for (int j = 0; j < size; j++)
        {
            if (j == colToRemove)
            {
                continue;
            }

            submatrix[rowIndex][colIndex] = matrix[i][j];
            colIndex++;
        }
        rowIndex++;
    }

    return submatrix;
}

int calculateDeterminant(vector<vector<int>> matrix)
{
    int size = matrix.size();

    if (size == 2)
    {
        return matrix[0][0] * matrix[1][1] - matrix[0][1] * matrix[1][0];
    }

    int determinant = 0;
    for (int i = 0; i < size; i++)
    {
        determinant += matrix[0][i] * pow(-1, i) * calculateDeterminant(getSubmatrix(matrix, 0, i));
    }

    return determinant;
}

vector<vector<int>> transposeMatrix(vector<vector<int>> matrix)
{
    int rows = matrix.size();
    int cols = matrix[0].size();
    vector<vector<int>> transposed(cols, vector<int>(cols, 0));

    for (int i = 0; i < rows; i++)
    {
        for (int j = 0; j < cols; j++)
        {
            transposed[j][i] = matrix[i][j];
        }
    }

    return transposed;
}

vector<vector<int>> calculateAdjoint(vector<vector<int>> matrix)
{
    int size = matrix.size();
    vector<vector<int>> adjoint(size, vector<int>(size, 0));

    if (size == 1)
    {
        adjoint[0][0] = 1;
    }
    else if (size == 2)
    {
        adjoint[0][0] = matrix[1][1];
        adjoint[0][1] = -matrix[0][1];
        adjoint[1][0] = -matrix[1][0];
        adjoint[1][1] = matrix[0][0];
    }
    else
    {
        for (int i = 0; i < size; i++)
        {
            for (int j = 0; j < size; j++)
            {
                vector<vector<int>> submatrix = getSubmatrix(matrix, i, j);
                adjoint[i][j] = pow(-1, i + j) * calculateDeterminant(submatrix);
            }
        }
        adjoint = transposeMatrix(adjoint);
    }

    return adjoint;
}

int modInverse(int A, int M)
{
    for (int i = 1; i < M; i++)
    {
        if ((((A % M) * (i % M)) % M) == 1)
        {
            return i;
        }
    }
    return -1;
}

vector<vector<int>> inverseOfKeyMatrix(vector<vector<int>> keyMatrix)
{
    int determinant = calculateDeterminant(keyMatrix);
    if (determinant < 0)
    {
        determinant = 26 - (abs(determinant) % 26);
    }
    determinant = determinant % 26;
    int modularInverse = modInverse(determinant, 26);

    vector<vector<int>> inverseMatrix = calculateAdjoint(keyMatrix);

    for (int i = 0; i < inverseMatrix.size(); i++)
    {
        for (int j = 0; j < inverseMatrix[0].size(); j++)
        {
            inverseMatrix[i][j] = (inverseMatrix[i][j]) % 26;
            if (inverseMatrix[i][j] < 0)
            {
                inverseMatrix[i][j] = 26 - abs(inverseMatrix[i][j]);
            }
            inverseMatrix[i][j] = (inverseMatrix[i][j] * modularInverse) % 26;
        }
    }
    return inverseMatrix;
}

void decrypt(string str, string key)
{
    if (!isPerfectSquare(key.length()))
    {
        cout << "Enter a key whose length is a perfect square root." << endl;
    }
    int root = (int)sqrt(key.length());
    vector<vector<int>> keyMat(root, vector<int>(root, 0));
    int k = 0;
    for (int i = 0; i < root; i++)
    {
        for (int j = 0; j < root; j++)
        {
            keyMat[i][j] = ((int)key[k] - 97) % 26;
            k++;
        }
    }
    int temp = 0;
    vector<vector<int>> inverseKey = inverseOfKeyMatrix(keyMat);
    vector<vector<int>> strMat(root, vector<int>(str.length() / root, 0));
    for (int i = 0; i < strMat[0].size(); i++)
    {
        for (int j = 0; j < strMat.size(); j++)
        {
            strMat[j][i] = ((int)str[temp] - 97) % 26;
            temp++;
        }
    }

    vector<vector<int>> mul = matrixMultiplication(inverseKey, strMat);
    string ans = "";
    for (int i = 0; i < mul[0].size(); i++)
    {
        for (int j = 0; j < mul.size(); j++)
            ans += (char)(mul[j][i] % 26 + 97);
    }

    cout << "Plain Text: " << ans << endl;
}